### PR TITLE
Upgrades to loads module

### PR DESCRIPTION
**Pull request by hivanov-nrel**
Hi guys,

I made some upgrades to the loads module that would be nice to assimilate into the master repo. The improvements/additions can be summarized below:

- improved accuracy of rainflow counting algorithm by increasing the load range division
- new function: _calculate_TSR__ - calculate the tip speed ratio of a MEC with a rotor.
- new function: _calculate_Cp_ - calculate the power coefficient (typically for MEC with a rotor).
- new function: _calculate_blade_moments_ - converts blade flap & edge strain gauge signals into engineering units.

These functions are all thrown into the general loads.py script for now, but some arent necessarily related to loads so I would be curious to suggestions where they might live. Please let me know your thoughts and if I need to make updates to be consistent with nomenclature and so forth.

### CODE REVIEW

File: mhkit/loads.py
Comment (ssolson) [inline mhkit/loads.py:205]: It looks like here that blade matrix is expected to have 4 parts. If this is so we should assert the dimension at the top and state so in the docstring.

File: mhkit/tests/test_loads.py
Comment (ssolson) [inline mhkit/tests/test_loads.py:61]: Will you consider using the built-in function [assertAlmostEqual](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertAlmostEqual) for this one and where appropriate below?

**Review by ssolson — COMMENTED** [review]
Hey, Chris thanks for putting this together and great job on this! I have a couple of small comments on the PR I would like to discuss.

File: mhkit/loads.py
Comment (hivanov-nrel) [reply mhkit/loads.py:85]: Should we say MEC or MHK? I chose MEC for now.

File: mhkit/loads.py
Comment (hivanov-nrel) [reply mhkit/loads.py:201]: Yes I am open to simplifying this. My only fear would be for someone to forget to apply it which could skew results without it being obvious. But this may be overcome by making it very apparent in the docstring that the input must already be corrected. @rpauly18 any thoughts?

File: mhkit/loads.py
Comment (hivanov-nrel) [reply mhkit/loads.py:168]: It is usually called a matrix but i thought that requesting an array/list is easier for the user to input.

File: mhkit/loads.py
Comment (rpauly18) [reply mhkit/loads.py:200]: I think we should leave the offsets as input because I have the same fears as Chris. Users will forget to apply the offsets. In my experience users do not read the docs that thoroughly, or even if they do, over time they may forget that they need to apply the offsets themselves. Adding a couple numbers to the function call will not be a heavy lift for the user.

File: mhkit/loads.py
Comment (rpauly18) [reply mhkit/loads.py:90]: I agree that unless RPM is the industry standard, we should use RPS for consistency. That being said, I do not have strong feelings about it.

**Comment by rpauly18** [comment]
@hivanov-nrel Great job on this. As a start before I give a more in depth review, I do think we are at point where we can do a re-org of the loads module since there are now more functions. Since we now have 2 functions that are blade/rotor type device specific we can do something like loads.general, loads.rotor, loads.graphics. I am not 100% sure those are the best names, but at the very least we should probably do a loads.graphics to maintain consistency with other modules. @ssolson  and @kmruehl what are your thoughts?

**Comment by kmruehl** [comment]
> 
> 
> @hivanov-nrel Great job on this. As a start before I give a more in depth review, I do think we are at point where we can do a re-org of the loads module since there are now more functions. Since we now have 2 functions that are blade/rotor type device specific we can do something like loads.general, loads.rotor, loads.graphics. I am not 100% sure those are the best names, but at the very least we should probably do a loads.graphics to maintain consistency with other modules. @ssolson and @kmruehl what are your thoughts?

I agree, and I'm in favor of keeping the names as similar to other modules as much as possible. Right now it's hard to say what the best breakdown should be since I don't have a sense of what the future development plans are, but they could include: `mhkit.loads.graphics`, `mhkit.loads.device` and `mhkit.loads.fatigue`

**Comment by ssolson** [comment]
I'm a big fan of breaking the loads module into submodules! I agree with the breakdown proposed by Rebecca.

**Comment by rpauly18** [comment]
@hivanov-nrel given that you know best what future development can be included in the loads module, what way do you think breaking down the module is best?

**Comment by hivanov-nrel** [comment]
Thanks everyone for the feedback! I will work on incorporating the rest of the comments in the next commit. I agree that a subdivision would be good for the loads module. The one concern I wanted to raise again was that some of the functions like ```tip_speed_ratio``` and ```power_coefficient``` are not necessarily related to loads. Typically, these quantities are used to check the power performance of a machine. Here is what I was thinking below about the subdivision. 
- ```loads.general``` 
   - ```bin_statistics```
   - ```blade_moments```
   - ```damage_equivalent_load```
- ```loads.graphics```
   - ```plot_bin_statistics```
   - ```plot_statistics```
- ```tidal.performance```
   - ```tip_speed_ratio```
   - ```power_coefficient```

I think the addition of a ```tidal.performance``` submodule would be good. I think functions that help assess the power performance of a device are definitely useful to a user. It looks like wave already has something like this. It would be great to break up the ```loads.general``` even more, but its hard to say how this module might evolve. I could see a ```loads.device``` submodule being populated based on different device needs, but I dont have anything besides the ```blade_moments``` for now. My current development plan for the future is to potentially add two more fatigue functions (cumulative_rainflow_counts, lifetime_damage_equivalent_load) and to try to make the loads examples more relatable by utilizing simulation data from WEC-Sim rather than using turbine data. My hope is that as more users start to use the modules, new functionalities will be requested that we can incorporate. Please let me know your thoughts.

File: mhkit/loads.py
Comment (hivanov-nrel) [reply mhkit/loads.py:90]: i will make it RPS for consistency, it should not be hard for the user to make the conversion if it is measured in RPM.

File: mhkit/tests/test_loads.py
Comment (hivanov-nrel) [reply mhkit/tests/test_loads.py]: updated the tests to use assertAlmostEqual

**Comment by kmruehl** [comment]
@hivanov-nrel I like your proposed breakdown, nice work!

**Comment by ssolson** [comment]
Chris I like the proposed break down. Please review the changes I submitted as a PR to your fork and incorporate into the new structure.

**Comment by rpauly18** [comment]
@hivanov-nrel your changes, including the changes to the init files look good. Thanks for your hard work on this!

**Review by rpauly18 — APPROVED** [review]

**Comment by ssolson** [comment]
Chris when you opened this PR you discussed that some of the functions were not necessarily related to loads. Now that we have broken out the functions are these the TSR and Cp calculations? What do you and the team think about these being in river/ tidal as "device" functions? I think they make more sense as part of the device functions over `loads.performance`
Were you thinking any of the other functions you wrote potentially belong in river/ tidal?

**Comment by rpauly18** [comment]
@ssolson the TSR and Cp functions are in tidal.performance, which I think is the appropriate location for them. The reason for using "performance" over "device" is to keep consistency with the breakdown used in wave. These functions, while they do describe the device, more specifically characterize the device performance. This does bring up the question of if we want to keep the current "device" modules as they are or merge them. 

TSR and Cp likely should be pulled onto the river module as well since river devices often use blade/rotor designs as well.

**Comment by ssolson** [comment]
Apologies I missed that they are in tidal. Is there a reason this was chosen over river? They need to be mirrored either way but currently, everything shared between the two modules is mirrored from river into tidal. I would vote to keep this consistent.

**Comment by rpauly18** [comment]
The tidal location was probably just because we have discussed the use of the loads module more in terms of tidal devices, not river, but at the end of the day it applies to both. If you prefer the functions be moved into river, I have no issues with that.

**Comment by ssolson** [comment]
Okay great, let's relocate it to river then. With respect to the device and performance module, I am not sure what to do. If you guys want to keep the performance module I would lean towards moving the IEC equivalent diameter calculations from river/tidal device into the new performance module. They are not really performance but this would bring consistency across the modules. Thoughts?

**Comment by rpauly18** [comment]
I think moving them into performance is appropriate since the IEC TS that is cited for that module is the power performance TS. Users should hopefully understand that connection.

**Comment by ssolson** [comment]
@hivanov-nrel check your fork's PRs for the new structure proposed above.

**Comment by hivanov-nrel** [comment]
@ssolson I have merged your PR. Thanks for making those updates.

**Comment by ssolson** [comment]
Chris/all I am going to give the PR a once over and then approve and merge if everything looks good.

**Comment by ssolson** [comment]
Okay Chris I created a PR ( just as a heads up I have been rejected from pushing directly to your fork) to address primarily an issue with not importing numpy into the graphics module and included some docstring fixes. 

Once you merge those changes I will approve the PR and merge. Thanks again!

**Review by ssolson — APPROVED** [review]

**Pull request merged**
